### PR TITLE
Adding HTTP Mutual Auth Demo from C-SDK

### PR DESCRIPTION
**Pull request by sukhmanm**
Description
-----------
Adding demo to demonstrate use of the coreHTTP library using a mutually-authenticated network connection. Sends a simple message using a POST request and waits for a server response.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: demos/coreHTTP/http_demo_mutual_auth.c [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+#include "transport_secure_sockets.h"
+
+/* Common HTTP demo utilities. */
+#include "http_demo_utils.h"
+
+/* HTTP API header. */
+#include "core_http_client.h"
+
+/* Include header for client credentials. */
+#include "aws_clientcredential_keys.h"
```
Comment (sarenameas) [inline demos/coreHTTP/http_demo_mutual_auth.c:73]: What do you need from this header in this file?

Comment (sarenameas) [reply demos/coreHTTP/http_demo_mutual_auth.c:73]: Should you include aws_clientcredential.h for the iot endpoint?

File: demos/coreHTTP/http_demo_mutual_auth.c [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+#endif
+
+/* Check that a request body to send for the POST request is defined. */
+#ifndef democonfigREQUEST_BODY
+    #error "Please define a democonfigREQUEST_BODY."
+#endif
+
+/* Check that the AWS IoT Core endpoint is defined. */
+#ifndef democonfigAWS_IOT_ENDPOINT
+    #define democonfigAWS_IOT_ENDPOINT    ""
```
Comment (sarenameas) [inline demos/coreHTTP/http_demo_mutual_auth.c:107]:
```suggestion
    #define democonfigAWS_IOT_ENDPOINT   clientcredentialMQTT_BROKER_ENDPOINT
```

File: demos/coreHTTP/http_demo_mutual_auth.c [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+static BaseType_t prvSendHttpRequest( const TransportInterface_t * pTransportInterface,
+                                      const char * pMethod,
+                                      size_t methodLen,
+                                      const char * pPath,
+                                      size_t pathLen );
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Entry point of demo.
```
Comment (sarenameas) [inline demos/coreHTTP/http_demo_mutual_auth.c:187]:
```suggestion
 * @brief Entry point of the demo.
```

File: demos/coreHTTP/http_demo_mutual_auth.c [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+    ( void ) awsIotMqttMode;
+    ( void ) pIdentifier;
+    ( void ) pNetworkServerInfo;
+    ( void ) pNetworkCredentialInfo;
+    ( void ) pNetworkInterface;
+
+    /**************************** Connect. ******************************/
+
+    /* Attempt to connect to the HTTP server. If connection fails, retry after
+     * a timeout. Timeout value will be exponentially increased till the maximum
```
Comment (sarenameas) [inline demos/coreHTTP/http_demo_mutual_auth.c:227]:
```suggestion
     * a timeout. The timeout value will be exponentially increased till the maximum
```

File: demos/coreHTTP/http_demo_mutual_auth.c [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+     * attempts are reached or maximum timeout value is reached. The function
+     * returns EXIT_FAILURE if the TCP connection cannot be established to
```
Comment (sarenameas) [inline demos/coreHTTP/http_demo_mutual_auth.c:229]:
```suggestion
     * returns pdFAIL if the TCP connection cannot be established to
```

File: demos/coreHTTP/http_demo_mutual_auth.c [hunk 6]
```diff
@@ -0,0 +1,444 @@
...
+     * broker after configured number of attempts. */
+    xDemoStatus = connectToServerWithBackoffRetries( prvConnectToServer,
+                                                     &xNetworkContext );
+
+    if( xDemoStatus == pdPASS )
+    {
+        /* Set a flag indicating a TLS connection exists. This is done to
+         * disconnect if the loop exits before disconnection happens. */
```
Comment (sarenameas) [inline demos/coreHTTP/http_demo_mutual_auth.c:237]: No more loop

File: demos/coreHTTP/http_demo_mutual_auth.c [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+    xSocketsConfig.maxFragmentLength = 0;
+    xSocketsConfig.disableSni = false;
+    xSocketsConfig.pRootCa = democonfigROOT_CA_PEM;
+    xSocketsConfig.rootCaSize = sizeof( democonfigROOT_CA_PEM );
+    xSocketsConfig.sendTimeoutMs = democonfigTRANSPORT_SEND_RECV_TIMEOUT_MS;
+    xSocketsConfig.recvTimeoutMs = democonfigTRANSPORT_SEND_RECV_TIMEOUT_MS;
+
+    /* Establish a TLS session with the HTTP server. This example connects to
+     * the HTTP server as specified in democonfigAWS_IOT_ENDPOINT and
+     * democonfigAWS_HTTP_PORT in demo_config.h. */
```
Comment (sarenameas) [inline demos/coreHTTP/http_demo_mutual_auth.c:316]:
```suggestion
     * democonfigAWS_HTTP_PORT in http_demo_mutual_auth_config.h. */
```

File: demos/coreHTTP/http_demo_mutual_auth.c (hunk 1 above)
Comment (sukhmanm) [reply demos/coreHTTP/http_demo_mutual_auth.c:73]: Nothing really -- removed it, thanks.

File: demos/coreHTTP/http_demo_mutual_auth.c (hunk 5 above)
Comment (sukhmanm) [reply demos/coreHTTP/http_demo_mutual_auth.c:229]: Good catch, thanks!

File: demos/common/http_demo_helpers/http_demo_utils.c [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    RetryUtilsParams_t xReconnectParams;
+
+    assert( connectFunction != NULL );
+
+    /* Initialize reconnect attempts and interval */
+    RetryUtils_ParamsReset( &xReconnectParams );
+    xReconnectParams.maxRetryAttempts = MAX_RETRY_ATTEMPTS;
+
+    /* Attempt to connect to HTTP server. If connection fails, retry after
+     * a timeout. Timeout value will exponentially increase until maximum
```
Comment (sarenameas) [inline demos/common/http_demo_helpers/http_demo_utils.c:49]:
```suggestion
     * a timeout. The timeout value will exponentially increase until the maximum
```

**Review by sarenameas — DISMISSED** [review]
If it passes on the 5 boards in CI

**Review by sarenameas — DISMISSED** [review]

**Review by sarenameas — DISMISSED** [review]

**Review by yourslab — APPROVED** [review]

**Review by cobusve — APPROVED** [review]

**Pull request merged**